workspace/lib/xod/core/i2c-read/arduino.cpp

```diff
@@ -0,0 +1,25 @@
+
+{{#global}}
+#include <Wire.h>
+{{/global}}
+
+struct State {
+};
+
+{{ GENERATED_CODE }}
+
+void evaluate(Context ctx) {
+    if (!isInputDirty<input_READ>(ctx))
+        return;
+
+    auto usLimit = micros() + getValue<input_TOUT>(ctx) * 1e6;
+    while (!Wire.available()) {
+        // TODO: we should not block here,
+        // but setTimeout(ctx, 0) to try again later
```

Comment:
Are there any difficulties to complete "TODO" now?

Reply:
Not so many.

The main one is a lack of research. It *could* be a problem if we don’t consume incoming I²C data quickly because of possible hardware overflow. Maybe it is not. The sensors I used for testing actually send the data right away and the timeout handling is not required for them at all.

A research on a module that is actually unstable/slow is required. I have no such right at this moment.